src/mlpack/methods/ann/layer/neural_turing_machine_impl.hpp

```diff
@@ -0,0 +1,406 @@
+      allZeros.n_rows, allZeros.n_cols, false, true)));
+
+  gReads = lReads.end();
+
+  forwardStep = 0;
+  backwardStep = 0;
+  gradientStep = 0;
+}
+
+template <typename InputDataType, typename OutputDataType>
+NeuralTuringMachine<InputDataType, OutputDataType>::~NeuralTuringMachine()
+{
+  boost::apply_visitor(deleteVisitor, readHead);
+  boost::apply_visitor(deleteVisitor, writeHead);
+  boost::apply_visitor(deleteVisitor, controller);
+  boost::apply_visitor(deleteVisitor, inputToLinear);
+  boost::apply_visitor(deleteVisitor, eraseGate);
+  boost::apply_visitor(deleteVisitor, addGate);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void NeuralTuringMachine<InputDataType, OutputDataType>::Forward(
+    arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  // Create input to the controller.
+  arma::vec newInput = arma::join_vert(input, lReads.back());
+
+  // Forward pass through the controller.
+  boost::apply_visitor(ForwardVisitor(std::move(newInput), std::move(
+      boost::apply_visitor(outputParameterVisitor, controller))),
+      controller);
+
+  // Get controller output.
+  arma::mat& controllerOutput = boost::apply_visitor(outputParameterVisitor,
+      controller);
+
+  const arma::mat& cMemory = memoryHistory.back();
+
+  // Pass the controller output through read head.
+  boost::apply_visitor(ForwardVisitor(std::move(controllerOutput),
+      std::move(boost::apply_visitor(outputParameterVisitor, readHead))),
+      readHead);
+  const arma::mat& readWeights = boost::apply_visitor(outputParameterVisitor,
+      readHead);
+
+  //! Read memory.
+  lReads.push_back(arma::trans(cMemory) * readWeights);
+
+  // Pass through linear layer.
+  boost::apply_visitor(ForwardVisitor(std::move(controllerOutput), std::move(
+      boost::apply_visitor(outputParameterVisitor, inputToLinear))),
+      inputToLinear);
+
+  arma::mat& lOutput = boost::apply_visitor(outputParameterVisitor,
+      inputToLinear);
+
+  // Generate AddVec
+  boost::apply_visitor(ForwardVisitor(std::move(lOutput.submat(0, 0,
+      memSize - 1, 0)), std::move(boost::apply_visitor(outputParameterVisitor,
+      addGate))), addGate);
+  const arma::mat& addVec = boost::apply_visitor(outputParameterVisitor,
+      addGate);
+
+  // Generate EraseVec
+  boost::apply_visitor(ForwardVisitor(std::move(lOutput.submat(memSize, 0,
+      2 * memSize - 1, 0)), std::move(boost::apply_visitor(
+      outputParameterVisitor, eraseGate))), eraseGate);
+  const arma::mat& eraseVec = boost::apply_visitor(outputParameterVisitor,
+      eraseGate);
+
+  // Generate write weights
+  boost::apply_visitor(ForwardVisitor(std::move(controllerOutput),
+      std::move(boost::apply_visitor(outputParameterVisitor, writeHead))),
+      writeHead);
+  const arma::mat& writeWeights = boost::apply_visitor(outputParameterVisitor,
+      writeHead);
+
+  arma::mat nMemory = cMemory;
+
+  auto addVecIt = addVec.begin();
```

Comment:
Do you mind to avoid ``auto`` here, I think it helps to understand the code if the type is immediately visible.